draft-irtf-cfrg-hpke.md

```diff
@@ -366,25 +367,25 @@ following way, where `Group` denotes the Diffie-Hellman group and
 
 ~~~
 def ExtractAndExpand(dh, kemContext):
-  prk = LabeledExtract(zero(0), "dh", dh)
-  return LabeledExpand(prk, "prk", kemContext, Nzz)
+  prk = LabeledExtract(zero(0), "GROUP-dh", dh)
+  return LabeledExpand(prk, "GROUP-prk", kemContext, Nzz)
```

Comment:
How necessary is the `GROUP` prefix? This complicates things a non-trivial but acceptable amount in implementation, since I'm trying to avoid allocation entirely.

Reply:
Calling @kwantam! My take is that this is just good hygiene. Without the prefix, it's harder to reason about separation of these different ROs.

Reply:
As @chris-wood says, it makes it easier to reasonable about random oracle separation. It may be overkill (I think in my original comment I said something like "this is probably paranoia, but...").

The kind of thing one might worry about without `GROUP` is a case where it's possible to confuse serialized elements of two different groups. In this case it *could* be possible to construct two exchanges with identical `kemContext` inputs. (Not clear that's enough to construct an attack, but we can completely rule out attacks of this kind by including `GROUP`.)

Reply:
I agree with @chris-wood and @kwantam that binding `zz` explicitly to the underlying DH is important.

Would it be easier for you to implement it avoiding allowation, @rozbb, if it was in the style of the context in KeySchedule? There, we include `kem_id`, `kdf_id`, `aead_id`, all as 2-byte fixed-sized values. Instead of doing "GROUP-dh", we could do:

```
  prk = LabeledExtract(zero(0), concat(kem_id, "dh"), dh)
  return LabeledExpand(prk, concat(kem_id, "prk"), kemContext, Nzz)
```

The `kem_id` already is unique per DHKEM, and also, it would avoid needing a separate table for the values of GROUP, because the table in the “Algorithm Identifiers” Section already has the `kem_id`s.

Reply:
Ok, actually, I suggest the following for Encap and Decap:

```
  kemContext = concat(I2OSP(kem_id, 2), enc, pkRm)
```

and this for AuthEncap and AuthDecap:

```
  kemContext = concat(I2OSP(kem_id, 2), enc, pkRm, pkSm)
```

Then it is exactly in the same style as in KeySchedule. The added label `kem_id` always has two bytes. I think the LabeledExtract call in ExtractAndExpand does not need to have `kem_id` as a label, because the intermediary value `dh` is never exposed. The important thing is that `zz` is explicitly bound to the scheme's name, and for that it is enough to have it in `kemContext`. (I think we had a similar reasoning when we decided what to include where in KeySchedule's context)

Reply:
On the other hand, binding to the GROUP plays a similar role as the RFCXXX label which we do include in *both* LabeledExtract *and* LabeledExpand, even if the LabeledExtract results are never exposes. And this is about, as @kwantam said in #108, that the calls are orthogonal, which makes it easier to reason with separate random oracles for separate ciphersuites.

So I have a tendency for including `kem_id` in both, as I suggested in my first reply in this thread.

What do you think?

Reply:
I like it! This has the same effect and is less text. Stand by for that change.

Reply:
@blipp fixed -- please have a look.

Reply:
Yes, this is much cleaner! In fact, DeriveKeyPair should be using `kem_id` as well, instead of the `desc` string. I'll make the PR after this lands.